src/upload/index.ts

```diff
@@ -0,0 +1,290 @@
+export type BlockRange = {
+  start: number;
+  end: number;
+};
+
+export type BlockFile = {
+  blocks: string;
+  manifest: string;
+  timestamp: number;
+};
+
+export class LightBlockUpload {
+  private cache: LightBlockCache;
+  private s3Client: S3Client;
+  private chunkSizeMb: number;
+  private bucket: string;
+  private latestPath = "latest.json";
+  private blockFileName = "blocks";
+
+  constructor(cache: LightBlockCache) {
+    this.cache = cache;
+    if (!process.env["BUCKET_ENDPOINT"]) {
+      throw new UploadError("BUCKET_ENDPOINT not set");
+    }
+    if (!process.env["BUCKET_ACCESS_KEY_ID"]) {
+      throw new UploadError("BUCKET_ACCESS_KEY_ID not set");
+    }
+    if (!process.env["BUCKET_SECRET_ACCESS_KEY"]) {
+      throw new UploadError("BUCKET_SECRET_ACCESS_KEY not set");
+    }
+    if (!process.env["BUCKET_NAME"]) {
+      throw new UploadError("BUCKET_NAME not set");
+    }
+    if (!process.env["UPLOAD_CHUNK_SIZE_MB"]) {
+      throw new UploadError("UPLOAD_CHUNK_SIZE_MB not set");
+    }
+
+    this.chunkSizeMb = parseInt(
+      process.env["UPLOAD_CHUNK_SIZE_MB"] as string,
+      10,
+    );
+    this.bucket = process.env["BUCKET_NAME"];
+    this.s3Client = new S3Client({
+      region: "auto",
+      endpoint: process.env["BUCKET_ENDPOINT"],
+      credentials: {
+        accessKeyId: process.env["BUCKET_ACCESS_KEY_ID"],
+        secretAccessKey: process.env["BUCKET_SECRET_ACCESS_KEY"],
+      },
+    });
+  }
+
+  async upload(): Promise<void> {
+    const latestJson = await this.getObject(this.latestPath);
+
+    let currentUploadSize = 0;
+    if (!latestJson) {
+      console.warn("No latest json, starting upload from beginning...");
+    } else {
+      const latest: BlockFile = JSON.parse(latestJson);
+      currentUploadSize = (await this.getFileSize(latest.blocks)) || 0;
+    }
+
+    try {
+      logger.info(
+        `Current file uploaded size: ${this.bytesToMbRounded(
+          currentUploadSize,
+        )} MB, creating new upload...`,
+      );
+      const files = await this.createBlockFiles(
+        this.blockFileName,
+        currentUploadSize,
+      );
+      const prefix = String(files.timestamp) + "/";
+
+      logger.info(`Upload: begin...`);
+
+      const uploadedBinary = await this.uploadFile(
+        prefix,
+        files.blocks,
+        "application/octet-stream",
+      );
+      logger.info(`Upload: binary file complete: ${uploadedBinary}`);
+
+      const gzipManifest = await this.gzipFile(
+        files.manifest,
+        `${files.manifest}.gz`,
+      );
+      const uploadedManifest = await this.uploadFile(
+        prefix,
+        gzipManifest,
+        "application/gzip",
+      );
+      logger.info(`Upload: manifest file complete: ${uploadedManifest}`);
+
+      const uploadedLatest = await this.writeLatestTimestamp(
+        uploadedManifest,
+        uploadedBinary,
+        files.timestamp,
+      );
+      await this.uploadFile("", uploadedLatest, "plain/text");
```

Comment:
Nit: shouldn't this be an `application/json`?

Reply:
Good call!